File: Heap/LC_973.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.12.05

class Solution {
public:
    vector<vector<int>> kClosest(vector<vector<int>>& points, int k) {
        //basic max heap based solution
        vector<vector<int>> ans;
        priority_queue<pair<int, vector<int>>, vector<pair<int, vector<int>>>> pq;
        for (int i = 0; i < points.size(); i++){
            int x = points[i][0], y = points[i][1];
            pq.push({x * x + y * y, {x, y}});
            if (pq.size() > k){
                pq.pop();
            }
        }
        while(!pq.empty()){
            auto item = pq.top();
            pq.pop();
            ans.push_back(item.second);
        }
        return ans;
        
    }
};

File: Heap/LC_973.py
# Written by: Jian-Yuan Lin
# Date: 2022.12.05

class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        maxheap = []
        for point in points:
            dist = -(point[0]**2 + point[1]**2)
            heapq.heappush(maxheap,(dist, point))
            if len(maxheap) > k:
                heapq.heappop(maxheap)
        ans = []
        for item in maxheap:
            ans.append(item[1])
        return ans